File: Leetcode/Top Interview Questions/46.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        answer = []
        def dfs(arr):
            if len(arr) == len(nums):
                answer.append(arr[:])
                return
            for num in nums:
                if num not in arr:
                    arr.append(num)
                    dfs(arr)
                    arr.pop()

        dfs([])
        return answer 
        

File: Leetcode/Recursion/46.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
  let answer = [];

  function makePermutations(curr, list) {
    if (curr.length === nums.length) return answer.push(curr);

    for (let i = 0; i < list.length; i++) {
      makePermutations(
        [...curr, list[i]],
        list.filter((num, j) => i !== j)
      );
    }
  }

  makePermutations([], nums);

  return answer;
};
